### PR TITLE
A bunch of fixes from static analysis and production

### PR DESCRIPTION
**Pull request by dimhotepus**

### CODE REVIEW

File: RtspServer/MediaTypes/RFC2250Media.cs [hunk 1]
```diff
@@ -284,8 +284,7 @@ public byte[] WriteMPEGSequence(Bitmap img)
 
                     //	Put memory stream (which contains the encoded image) into buffer
                     ms = MPEG.getMS();
-                    byte[] buffer = new Byte[ms.Length];
-                    buffer = ms.ToArray();
+                    byte[] buffer = ms.ToArray();
```
Comment (juliusfriedman) [inline RtspServer/MediaTypes/RFC2250Media.cs]: I think ToArray will still allocate here (https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/IO/MemoryStream.cs#L568), TryGetBuffer is guaranteed to not allocate but I can fix this later.

File: RtspServer/RtspStreamArchiver.cs [hunk 2]
```diff
@@ -57,60 +57,56 @@ public virtual void WritePacket(IMedia stream, Common.IPacket packet)
...
 
             }
         }
 
         void RtpClientPacketReceieved(object sender, Common.IPacket packet = null, Media.Rtp.RtpClient.TransportContext tc = null)
         {
-            if(sender is Rtp.RtpClient)
-                WritePacket(Attached.Keys.FirstOrDefault(s => (s as RtpSource).RtpClient == sender as Rtp.RtpClient), packet);
+            if(sender is Rtp.RtpClient c)
+                WritePacket(Attached.Keys.FirstOrDefault(s => (s as RtpSource).RtpClient == c), packet);
```
Comment (juliusfriedman) [inline RtspServer/RtspStreamArchiver.cs]: I think this should ```s is RtpSource source && source.RtpClient == c```.

File: RtspServer/RtspStreamArchiver.cs [hunk 3]
```diff
@@ -138,7 +133,7 @@ public ArchiveSource(string name, Uri source, Guid id)
             }
             
 
-            public List<RtpSource> Playback = new List<RtpSource>();
+            public List<RtpSource> Playback = new();
```
Comment (juliusfriedman) [inline RtspServer/RtspStreamArchiver.cs]: readonly?

File: Common/Extensions/OperatingSystemExtensions.cs
Comment (juliusfriedman) [inline Common/Extensions/OperatingSystemExtensions.cs:1]: Thank  you!

File: Common/Classes/TaggedException.cs [hunk 4]
```diff
@@ -151,6 +153,7 @@ public TaggedException(T tag, string message, System.Exception innerException)
         /// <param name="tag"></param>
         /// <param name="info"><see cref="System.Runtime.Serialization.SerializationInfo"/></param>
         /// <param name="context"><see cref="System.Runtime.Serialization.StreamingContext"/></param>
+        [Obsolete]
```
Comment (juliusfriedman) [inline Common/Classes/TaggedException.cs:156]: 👍 It served it purpose well, if its going to be obsoleted we should probably provide a small guidance comment on how to throw and exception and catch it with user supplied data or make such an example used in the library and refer to it in the comment so there is less confusion.

File: Common/Extensions/ExceptionExtensions.cs [hunk 5]
```diff
@@ -49,20 +49,6 @@ public static class ExceptionExtensions
         /// </summary>
         public const System.Exception NilException = null;
 
-        /// <summary>
-        /// Check if we are in a exception unwind scenario or not.
-        /// </summary>
-        public static bool InException
```
Comment (juliusfriedman) [inline Common/Extensions/ExceptionExtensions.cs:55]: I think Obsolete here also would help, this was never really used but it used to serve the purpose of `how to check if code is in a throw state`

File: RtspServer/MediaTypes/RFC2250Media.cs (hunk 1 above)
Comment (juliusfriedman) [reply RtspServer/MediaTypes/RFC2250Media.cs]: I believe a lot of these allocations can still be reduced also but its a larger task. Lets go with something simple like this which reduce a majority of the duplicate allocations.
```
public ArraySegment<byte> WriteMPEGSequence(Bitmap img)
{
    int i;
    int j, k1, k2;
    long j2;
    byte tempByte;

    int ACSIZE = 1764;
    byte[] leftoverBits = new byte[10];
    byte[] DCbits = new byte[24];
    byte[] ACbits = new byte[ACSIZE];

    int DCY, DCCR, DCCB, lastDCY, lastDCCR, lastDCCB;
    int hblock, vblock;
    byte[,] Y = new byte[16, 16];
    byte[,] CR = new byte[8, 8];
    byte[,] CB = new byte[8, 8];
    byte[,] block = new byte[8, 8];
    double[,] S = new double[8, 8];
    int[,] Q = new int[8, 8];
    int[] ZZ = new int[64];
    long imageBytes = 0;
    long bitRate;

    MPEGFunctions MPEG = new MPEGFunctions();

    imageBytes = img.Height * img.Width * 3;

    using (MemoryStream wms = new MemoryStream())
    {
        // Create output file and Memory Stream to write encoded image to
        using (BinaryWriter bw = new BinaryWriter(wms))
        {
            MemoryStream ms = new MemoryStream();


            //	Set up variables to encode image into MPEG frame
            lastDCY = 128;
            lastDCCR = lastDCCB = 128;
            for (i = 0; i < 10; i++)
                leftoverBits[i] = 255;
            for (i = 0; i < 24; i++)
                ACbits[i] = 255;
            for (i = 0; i < 24; i++)
                DCbits[i] = 255;

            long outBytes = 20;

            // Write MPEG picture and slice headers to MemoryStream
            for (i = 0; i < 10; i++)
                MPEG.picHeaderBits[i + 32] = (byte)((0 & (int)Math.Pow(2, 9 - i)) >> (9 - i));
            MPEG.writeToMS(leftoverBits, MPEG.picHeaderBits, ACbits, ref outBytes);
            MPEG.writeToMS(leftoverBits, MPEG.sliceHeaderBits, ACbits, ref outBytes);

            // Do this for each 16x16 pixel block in the bitmap file
            for (vblock = 0; vblock < img.Height / 16; vblock++)
                for (hblock = 0; hblock < img.Width / 16; hblock++)
                {
                    //	Write 2 bits for Macroblock header to leftoverbits
                    //	leftoverbits = '1', '1';
                    MPEG.writeMbHeader(leftoverBits);

                    //	Fill the Y[] array with a 16x16 block of RGB values
                    Y = MPEG.getYMatrix(img, vblock, hblock);
                    //	Fill the CR and CB arrays with 8x8 blocks by subsampling 
                    //	the RGB array
                    CR = MPEG.getCRMatrix(img, vblock, hblock);
                    CB = MPEG.getCBMatrix(img, vblock, hblock);

                    // First calculate DCTs for the 4 Y blocks
                    for (k1 = 0; k1 < 2; k1++)
                        for (k2 = 0; k2 < 2; k2++)
                        {
                            //	Put 8x8 Y blocks into the block[] array and
                            //	then calculate the DCT and quantize the result
                            for (i = 0; i < 8; i++)
                                for (j = 0; j < 8; j++)
                                    block[i, j] = Y[(k1 * 8 + i), (k2 * 8 + j)];
                            S = MPEG.calculateDCT(block);
                            Q = MPEG.Quantize(S);

                            //	Section to differentially Huffman encode DC values
                            //	DC is the diffential value for the DC coefficient
                            //	lastDC is the running total of the full magnitude
                            //	Then send the DC value to DCHuffmanEncode
                            for (i = 0; i < 24; i++)
                                DCbits[i] = 255;
                            DCY = Q[0, 0] - lastDCY;
                            lastDCY += DCY;
                            DCbits = MPEG.DCHuffmanEncode(DCY, MPEG.DCLumCode, MPEG.DCLumSize);

                            //	Section to encode AC Huffman values
                            //	Put the AC coefficients into the ACarray[]
                            //	in zigzag order, then Huffman encode the
                            //	resulting array.
                            for (i = 0; i < ACSIZE; i++)
                                ACbits[i] = 255;
                            ZZ = MPEG.Zigzag(Q);
                            ACbits = MPEG.ACHuffmanEncode(ZZ);

                            //	Write the encoded bits to the MemoryStream
                            MPEG.writeToMS(leftoverBits, DCbits, ACbits, ref outBytes);
                        }

                    // Now calculate the DCT for the CB array and quantize
                    S = MPEG.calculateDCT(CB);
                    Q = MPEG.Quantize(S);

                    //	Encode DC value
                    for (i = 0; i < 24; i++)
                        DCbits[i] = 255;
                    DCCB = Q[0, 0] - lastDCCB;
                    lastDCCB += DCCB;
                    DCbits = MPEG.DCHuffmanEncode(DCCB, MPEG.DCChromCode, MPEG.DCChromSize);

                    //	Encode AC values
                    for (i = 0; i < ACSIZE; i++)
                        ACbits[i] = 255;
                    ZZ = MPEG.Zigzag(Q);
                    ACbits = MPEG.ACHuffmanEncode(ZZ);

                    //	Write the encoded bits to the MemoryStream
                    MPEG.writeToMS(leftoverBits, DCbits, ACbits, ref outBytes);

                    // Now calculate the DCT for the CR array and quantize
                    S = MPEG.calculateDCT(CR);
                    Q = MPEG.Quantize(S);

                    // Encode DC value
                    for (i = 0; i < 24; i++)
                        DCbits[i] = 255;
                    DCCR = Q[0, 0] - lastDCCR;
                    lastDCCR += DCCR;
                    DCbits = MPEG.DCHuffmanEncode(DCCR, MPEG.DCChromCode, MPEG.DCChromSize);

                    //	Encode AC values
                    for (i = 0; i < ACSIZE; i++)
                        ACbits[i] = 255;
                    ZZ = MPEG.Zigzag(Q);
                    ACbits = MPEG.ACHuffmanEncode(ZZ);

                    //	Write the encoded bits to the MemoryStream
                    MPEG.writeToMS(leftoverBits, DCbits, ACbits, ref outBytes);
                }

            // Write EOP bits to the MemoryStream
            MPEG.writeEOP(leftoverBits, MPEG.EOPBits);
            outBytes++;

            //	Put memory stream (which contains the encoded image) into buffer
            ms = MPEG.getMS();

            if (!ms.TryGetBuffer(out var buffer)) buffer = ms.ToArray();

            //	Set MPEG Sequence Header bits to correct image size
            j = 2048;
            for (i = 0; i < 12; i++)
            {
                MPEG.seqHeaderBits[i + 32] = (byte)((j & img.Width) >> (11 - i));
                MPEG.seqHeaderBits[i + 44] = (byte)((j & img.Height) >> (11 - i));
                j >>= 1;
            }

            //	Set MPEG Sequence Header bits to bitRate value
            bitRate = ms.Length * 30 * 8 / 400;
            j2 = 131072;
            for (i = 0; i < 18; i++)
            {
                MPEG.seqHeaderBits[i + 64] = (byte)((j2 & bitRate) >> (17 - i));
                j2 >>= 1;
            }

            //	Write MPEG Sequence header to file
            for (i = 0; i < 12; i++)
            {
                tempByte = 0;
                for (j = 0; j < 8; j++)
                    tempByte = (byte)(tempByte * 2 + MPEG.seqHeaderBits[i * 8 + j]);
                bw.Write(tempByte);
            }

            //	Write MPEG GOP header to file
            for (i = 0; i < 8; i++)
            {
                tempByte = 0;
                for (j = 0; j < 8; j++)
                    tempByte = (byte)(tempByte * 2 + MPEG.GOPHeaderBits[i * 8 + j]);
                bw.Write(tempByte);
            }

            //	Fix the picture header for each MPEG frame and write 
            //	the buffer to the file
            for (i = 0; i < 1; i++)
            {
                for (j = 0; j < 10; j++)
                    MPEG.picHeaderBits[j + 32] = (byte)((i & (int)Math.Pow(2, 9 - j)) >> (9 - j));
                for (j = 0; j < 4; j++)
                {
                    tempByte = 0;
                    for (k1 = 0; k1 < 8; k1++)
                        tempByte = (byte)(2 * tempByte + MPEG.picHeaderBits[j * 8 + k1]);
                    buffer[j] = tempByte;
                }
                bw.Write(buffer);
            }

            // Write the End Of Sequence header
            bw.Write((byte)0x00);
            bw.Write((byte)0x00);
            bw.Write((byte)0x01);
            bw.Write((byte)0xb7);

            bw.Close();
            return wms.TryGetBuffer(out var result) ? result : null;
        }
    }
}
```

File: RtspServer/RtspServer.cs [hunk 6]
```diff
@@ -2828,7 +2828,7 @@ header field.
 
                     //Should use auth-int and qop
 
-                    authenticateHeader = string.Format(System.Globalization.CultureInfo.InvariantCulture, "Digest username={0},realm={1},nonce={2},cnonce={3}", requiredCredential.UserName, (string.IsNullOrWhiteSpace(requiredCredential.Domain) ? ServerName : requiredCredential.Domain), ((long)(Utility.Random.Next(int.MaxValue) << 32 | (Utility.Random.Next(int.MaxValue)))).ToString("X"), Utility.Random.Next(int.MaxValue).ToString("X"));
+                    authenticateHeader = string.Format(System.Globalization.CultureInfo.InvariantCulture, "Digest username={0},realm={1},nonce={2},cnonce={3}", requiredCredential.UserName, (string.IsNullOrWhiteSpace(requiredCredential.Domain) ? ServerName : requiredCredential.Domain), (((long)Utility.Random.Next(int.MaxValue)) << 32 | (long)(Utility.Random.Next(int.MaxValue))).ToString("X"), Utility.Random.Next(int.MaxValue).ToString("X"));
```
Comment (juliusfriedman) [inline RtspServer/RtspServer.cs:2831]: Good catch

File: RtspServer/MediaTypes/RFC6295Media.cs [hunk 7]
```diff
@@ -39,7 +39,7 @@ public class RFC6295Frame : Rtp.RtpFrame
             /// <returns></returns>
             internal static byte[] GetDeltaTime(uint delta)
             {
-                if (delta <= 0x0FFFFFFF) return Media.Common.MemorySegment.EmptyBytes;
+                if (delta >= 0x0FFFFFFF) return Media.Common.MemorySegment.EmptyBytes;
```
Comment (juliusfriedman) [inline RtspServer/MediaTypes/RFC6295Media.cs:42]: Good catch

File: RtspServer/RtspServer.csproj [hunk 8]
```diff
@@ -7,7 +7,7 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="System.Drawing.Common" Version="7.0.0" />
+    <PackageReference Include="System.Drawing.Common" Version="8.0.0" />
   </ItemGroup>
```
Comment (juliusfriedman) [inline RtspServer/RtspServer.csproj:11]: Thank you

File: RtspServer/MediaTypes/RtpVideoSink.cs [hunk 9]
```diff
@@ -2,12 +2,11 @@
...
 using Media.Common.Extensions.TimeSpan;
 using Media.Rtp;
-using Media.Rtsp.Server.MediaTypes;
-using Media.Sdp.Lines;
 using Media.Sdp;
+using Media.Sdp.Lines;
 using System;
-using static Media.Rtsp.Server.MediaTypes.RFC2435Media;
 using System.Collections.Generic;
+using static Media.Rtsp.Server.MediaTypes.RFC2435Media;
```
Comment (juliusfriedman) [inline RtspServer/MediaTypes/RtpVideoSink.cs]: The goal is eventually to remove RFC2435 Media from having to be used in a all the sinks, Lets just remove it now and remove this logic in SendPackets: 

```
                        
//Check for if previews should be updated (only for the jpeg type for now)
if (DecodeFrames && frame.PayloadType == RFC2435Frame.RtpJpegPayloadType)
{
    OnFrameDecoded((RFC2435Media.RFC2435Frame)frame);
}
```

File: RtspServer/MediaTypes/RtpVideoSink.cs [hunk 10]
```diff
@@ -197,20 +199,21 @@ internal override void SendPackets()
...
 
-                        //Take all the packet from the frame                            
+                        //Take all the packet from the frame
                         IEnumerable<Rtp.RtpPacket> packets = frame;
 
                         //Clear the frame to reset sequence numbers (could add method to do this)
                         //frame.RemoveAllPackets();
 
-                        if (Loop) frame = new Rtp.RtpFrame();
+                        if (Loop) frame = [];
```
Comment (juliusfriedman) [inline RtspServer/MediaTypes/RtpVideoSink.cs:215]: Collection Initializers , Lovely how it just looks like a plain array but its actually newing up a RtpFrame 👍

**Review by juliusfriedman — CHANGES_REQUESTED** [review]
Overall excellent work, just a few small changes here and there.

Thank you for your work!

File: RtspServer/MediaTypes/RFC2435Media.cs [hunk 11]
```diff
@@ -2317,15 +2317,8 @@ public RFC2435Media(string name, string directory, bool watch, int width, int he
...
-            int over;
-
-            Math.DivRem(Width, Common.Binary.BitsPerByte, out over);
-
-            /*if (over > 0) */Width += over;
-
-            Math.DivRem(Height, Common.Binary.BitsPerByte, out over);
-
-            /*if (over > 0) */Height += over;
+            Width += Width % Common.Binary.BitsPerByte;
```
Comment (juliusfriedman) [inline RtspServer/MediaTypes/RFC2435Media.cs:2320]: It used to be hard to get DivRem to be called but this still calls it per sharplab..

Thank you!

https://sharplab.io/#v2:C4LghgzgtgNAJiA1AHwAICYCMBYAUKgZgAIMiBhIgbzyNqIEsA7YIgdXrmAAsYiAJAKb0A5l2ABuGnUIkALEQCyACgCUVKXU3tOXIogC8bDtyIBSIgA5JuTbcEixew/dEtzVjUQC+nzzNTyAHKqntQ2trRMLAD2AG4CAE7WnhG0CmDcAHQAIvSxAEoCUEra3LwWvNEArjHxCSrJ4al0APQAVPQAZkRKcYlEAHxEAAxqbS2lugZEfUl4KanpWbkFRUouYuWVNTN1DfNNzUTtXT2zgyNjLRss07PWmj64XkA==

File: Media/Media.Consolidated.csproj
Comment (juliusfriedman) [inline Media/Media.Consolidated.csproj:1]: This file existed for someone who was doing something like a library they wanted to be able to update in a single shot the resulting binary without having to build and update each dll in the library.

I personally don't care that we removed it but I wanted to make sure it's intentional.

Comment (dimhotepus) [reply Media/Media.Consolidated.csproj:1]: Yep, removed intentionally as has no useful stuff. Only prints `Hello world!` in console.

**Review by juliusfriedman — APPROVED** [review]
LGTM, Thank you very much for your continued contributions.

**Comment by juliusfriedman** [comment]
Approved and merged!!!

**Pull request merged**